File: Editorial/320_Generalized_Abbreviation/CPP/1_Backtracking.cpp
class Solution {
public:
    void storeAbbreviations(vector<string>& abbreviations, string& word,
                            string currWord, int index, int abbreviatedCount) {
        if (index == word.size()) {
            // If the length of the last abbreviated substring is 0, add an
            // empty string.
            currWord +=
                abbreviatedCount == 0 ? "" : to_string(abbreviatedCount);

            abbreviations.push_back(currWord);
            return;
        }

        // Abbreviated string corresponding to the last substring chosen.
        string abbreviatedString =
            abbreviatedCount == 0 ? "" : to_string(abbreviatedCount);

        storeAbbreviations(abbreviations, word,
                           currWord + abbreviatedString + word[index],
                           index + 1, 0);
        storeAbbreviations(abbreviations, word, currWord, index + 1,
                           abbreviatedCount + 1);
    }

    vector<string> generateAbbreviations(string word) {
        vector<string> abbreviations;

        storeAbbreviations(abbreviations, word, "", 0, 0);
        return abbreviations;
    }
};

File: Editorial/0301-0400/0320_Generalized_Abbreviation/Java/1_Backtracking.java
class Solution {

    private void storeAbbreviations(
        List<String> abbreviations,
        String word,
        StringBuilder currWord,
        int index,
        int abbreviatedCount
    ) {
        if (index == word.length()) {
            // If the length of the last abbreviated substring is 0, add an empty string.
            if (abbreviatedCount > 0) {
                currWord.append(abbreviatedCount);
            }
            abbreviations.add(currWord.toString());
            return;
        }
        int currWordLength = currWord.length();
        // Option 1: Keep the current character.
        if (abbreviatedCount > 0) {
            currWord.append(abbreviatedCount);
        }
        currWord.append(word.charAt(index));
        storeAbbreviations(abbreviations, word, currWord, index + 1, 0);
        currWord.setLength(currWordLength); // Backtrack
        // Option 2: Abbreviate the current character.
        storeAbbreviations(
            abbreviations,
            word,
            currWord,
            index + 1,
            abbreviatedCount + 1
        );
    }

    public List<String> generateAbbreviations(String word) {
        List<String> abbreviations = new ArrayList<>();
        storeAbbreviations(abbreviations, word, new StringBuilder(), 0, 0);
        return abbreviations;
    }
}
